identity-account/src/account/account.rs

```diff
@@ -255,166 +252,151 @@ impl Account {
-    let method: &TinyMethod = old_state.capability_invocation()?;
-    let location: &KeyLocation = method.location();
+    let mut new_doc: IotaDocument = new_state.document().to_owned();
 
-    old_state
-      .sign_data(self.did(), self.storage(), location, &mut diff)
+    new_doc.set_previous_message_id(*self.chain_state().last_integration_message_id());
+
+    self
+      .sign_self(old_state.unwrap_or(new_state), new_state, &mut new_doc)
       .await?;
 
-    let message: MessageId = if self.config.testmode {
-      MessageId::null()
+    log::debug!(
+      "[publish_integration_change] publishing on index {}",
+      new_doc.integration_index()
+    );
+
+    let message_id: MessageId = if self.config.testmode {
+      // Fake publishing by returning a random message id.
+      MessageId::new(unsafe { crypto::utils::rand::gen::<[u8; 32]>().unwrap() })
     } else {
-      self
-        .client_map
-        .publish_diff(old_state.this_message_id(), &diff)
-        .await?
-        .into()
+      self.client_map.publish_document(&new_doc).await?.into()
     };
 
-    let events: [Event; 1] = [Event::new(EventData::DiffMessage(message))];
-
-    Self::commit_events(self.did(), &self.config, self.storage(), &new_root, &events).await?;
+    self.chain_state.set_last_integration_message_id(message_id);
 
     Ok(())
   }
 
-  async fn fold_snapshot(snapshot: IdentitySnapshot, commit: Commit) -> Result<IdentitySnapshot> {
-    Ok(IdentitySnapshot {
-      sequence: commit.sequence().max(snapshot.sequence),
-      identity: commit.into_event().apply(snapshot.identity).await?,
-    })
-  }
-
-  #[doc(hidden)]
-  pub async fn load_snapshot(&self) -> Result<IdentitySnapshot> {
-    // Retrieve the state snapshot from storage or create a new one.
-    let initial: IdentitySnapshot = self
-      .storage()
-      .snapshot(self.did())
-      .await?
-      .unwrap_or_else(|| IdentitySnapshot::new(IdentityState::new()));
-
-    // Apply all recent events to the state and create a new snapshot
-    self
-      .storage()
-      .stream(self.did(), initial.sequence())
-      .await?
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+  async fn publish_diff_change(&mut self, old_state: &IdentityState) -> Result<()> {
+    log::debug!("[publish_diff_change] publishing {:?}", self.document().did());
 
-  async fn load_snapshot_at(&self, generation: Generation) -> Result<IdentitySnapshot> {
-    let initial: IdentitySnapshot = IdentitySnapshot::new(IdentityState::new());
+    let old_doc: &IotaDocument = old_state.document();
+    let new_doc: &IotaDocument = self.state().document();
 
-    // Apply all events up to `generation`
-    self
-      .storage()
-      .stream(self.did(), Generation::new())
-      .await?
-      .take(generation.to_u32() as usize)
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+    let mut diff_id: &MessageId = self.chain_state().last_diff_message_id();
```

Comment:
Nit: maybe `previous_message_id` or `diff_previous_message_id` would be more descriptive? Completely optional.